Mathlib/Order/OrderType.lean

```diff
@@ -0,0 +1,450 @@
+   (Prod.Lex.uniqueProd β α).trans (Prod.Lex.prodUnique α β).symm
+
+open Classical in
+lemma mul_one (o : OrderType) : o * 1 = o :=
+  inductionOn o (fun α _ ↦ RelIso.ordertype_congr (Prod.Lex.prodUnique α PUnit))
+
+open Classical in
+lemma one_mul (o : OrderType) : 1 * o = o :=
+  inductionOn o (fun α _ ↦ RelIso.ordertype_congr (Prod.Lex.uniqueProd PUnit α))
+
+
+/-- `Equiv.prodAssoc` promoted to an order isomorphism. -/
+@[simps! (attr := grind =)]
+def OrderIso.prodAssoc (α : Type u) (β : Type v) (γ : Type w) [LE α] [LE β] [LE γ] :
+    (α × β) × γ ≃o α × (β × γ) :=
+  { Equiv.prodAssoc α β γ with
+    map_rel_iff' := fun {a b} ↦ by
+      rcases a with ⟨⟨_ , _⟩ , _⟩ ; rcases b with ⟨⟨_, _⟩ , _⟩ ;
+      simp [Equiv.prodAssoc, and_assoc] }
+
+/-- `Equiv.prodAssoc` promoted to an order isomorphism of lexicographic products. -/
+def OrderIso.prodLexAssoc (α : Type u) (β : Type v) (γ : Type w)
+    [LinearOrder α] [LinearOrder β] [LinearOrder γ] : (α ×ₗ β) ×ₗ γ ≃o α ×ₗ β ×ₗ γ :=
+  { OrderIso.prodAssoc α β γ with
+    map_rel_iff' := fun {a b} ↦
+      ⟨fun h ↦
+        match a, b, h with
+        | ⟨⟨a1 , a2⟩ , a3⟩ , ⟨⟨b1, b2⟩ , b3⟩ , h1 => by
+          rw [Prod.Lex.le_iff] at *
+          sorry --should be so simple
+        ,
+        fun h ↦
+        match a, b, h with
+        | ⟨⟨_ , _⟩ , _⟩ , ⟨⟨_, _⟩ , _⟩ , h1 => by sorry⟩}
+
+
+/-- `Equiv.prodSumDistrib` promoted to an order isomorphism of lexicographic products. -/
+def OrderIso.prodSumDistrib (α : Type u) (β : Type v) (γ : Type w)
+    [LinearOrder α] [LinearOrder β] [LinearOrder γ] : α ×ₗ (β ⊕ₗ γ) ≃o (α ×ₗ β) ⊕ₗ (α ×ₗ γ) :=
+  { Equiv.prodSumDistrib α β γ with
+    map_rel_iff' := fun {a b} ↦
+      ⟨fun h ↦
+        match a, b, h with
+        | ⟨a1 , (Sum.inlₗ a2)⟩ ,⟨b1 , (Sum.inlₗ b2)⟩ , h => by sorry
+        | ⟨a1 , (Sum.inrₗ a2)⟩ ,⟨b1 , (Sum.inlₗ b2)⟩ , h => by sorry
+        | ⟨a1 , (Sum.inlₗ a2)⟩ ,⟨b1 , (Sum.inrₗ b2)⟩ , h => by sorry
+        | ⟨a1 , (Sum.inrₗ a2)⟩ ,⟨b1 , (Sum.inrₗ b2)⟩ , h => by sorry,
+        fun h ↦
+        match a, b, h with
+        | ⟨a1 , (Sum.inlₗ a2)⟩ ,⟨b1 , (Sum.inlₗ b2)⟩ , h => by sorry
+        | ⟨a1 , (Sum.inrₗ a2)⟩ ,⟨b1 , (Sum.inlₗ b2)⟩ , h => by sorry
+        | ⟨a1 , (Sum.inlₗ a2)⟩ ,⟨b1 , (Sum.inrₗ b2)⟩ , h => by sorry
+        | ⟨a1 , (Sum.inrₗ a2)⟩ ,⟨b1 , (Sum.inrₗ b2)⟩ , h => by sorry⟩ }
+
+open Classical in
+lemma mul_assoc (o₁ o₂ o₃ : OrderType.{u}) : o₁ * o₂ * o₃ = o₁ * (o₂ * o₃) :=
+  inductionOn₃ o₁ o₂ o₃ (fun α _ β _ γ _ ↦ RelIso.ordertype_congr (OrderIso.prodLexAssoc α β γ))
+
+instance : Monoid OrderType where
+  mul_assoc := mul_assoc
+  one_mul := one_mul
+  mul_one := mul_one
+
+instance : LeftDistribClass OrderType where
+  left_distrib := fun o₁ o₂ o₃ ↦
+    inductionOn₃ o₁ o₂ o₃ (fun α _ β _ γ _ ↦ RelIso.ordertype_congr (OrderIso.prodSumDistrib α β γ))
+
+section Ordinal
+
+def LinearOrder.toWellOrder (α : Type u) [LinearOrder α] [IsWellOrder α (· < ·)] : WellOrder :=
```

Comment:
a) This should go on another file.
b) I don't think we even want this. We should eventually just redefine ordinals as a subtype of `OrderType`.